src/hook.rs

```diff
@@ -0,0 +1,162 @@
+use std::borrow::Cow;
+use libc::{self, c_int};
+use ffi::{self, lua_State, lua_Debug};
+use lua::extra_data;
+use util::callback_error;
+
+/// Contains information about the running code at the moments specified when setting the hook.
+/// Each field is described in the [Lua 5.3 documentaton][lua_doc].
+///
+/// Depending on when the hook is called, some fields might not be available. In these cases,
+/// integers and booleans might not be valid and/or strings are set to `None`.
+///
+/// [lua_doc]: https://www.lua.org/manual/5.3/manual.html#lua_Debug
+#[derive(Clone, Debug)]
+pub struct Debug<'a> {
+    pub name: Option<Cow<'a, str>>,
+    pub namewhat: Option<Cow<'a, str>>,
+    pub what: Option<Cow<'a, str>>,
+    pub source: Option<Cow<'a, str>>,
+    pub curr_line: u32,
+    pub line_defined: u32,
+    pub last_line_defined: u32,
+    pub num_ups: u32,
+    pub num_params: u32,
+    pub is_vararg: bool,
+    pub is_tailcall: bool,
+    pub short_src: Option<Cow<'a, str>>,
+
+    #[doc(hidden)]
+    _unused: ()
+}
+
+impl<'a> Debug<'a> {
+    /// Construct a new `Debug` structure that is not associated with a Lua debug structure. It
+    /// involves some string copying.
+    pub fn to_owned(&'a self) -> Debug<'static> {
+        Debug {
+            name: self.name.as_ref().and_then(|s| Some(Cow::Owned(s.as_ref().to_string()))),
+            namewhat: self.namewhat.as_ref().and_then(|s| Some(Cow::Owned(s.as_ref().to_string()))),
+            what: self.what.as_ref().and_then(|s| Some(Cow::Owned(s.as_ref().to_string()))),
+            source: self.source.as_ref().and_then(|s| Some(Cow::Owned(s.as_ref().to_string()))),
+            curr_line: self.curr_line,
+            line_defined: self.line_defined,
+            last_line_defined: self.last_line_defined,
+            num_ups: self.num_ups,
+            num_params: self.num_params,
+            is_vararg: self.is_vararg,
+            is_tailcall: self.is_tailcall,
+            short_src: self.short_src.as_ref().and_then(|s| Some(Cow::Owned(s.as_ref().to_string()))),
+            _unused: ()
+        }
+    }
+}
+
+/// Indicate in which circumstances the hook should be called by Lua.
+///
+/// # Usage
+///
+/// In order to clearly show which fields you are setting to `true` or `Some`, it is highly
+/// recommended you use the `Default` trait to fill in any remaining fields. This will also cover
+/// you in case new hook features gets added to Lua.
+///
+/// # Example
+///
+/// Constructs a `HookTriggers` structure that tells Lua to call the hook after every instruction.
+/// Note the use of `Default`.
+///
+/// ```
+/// # use rlua::HookTriggers;
+/// # fn main() {
+/// let triggers = HookTriggers {
+///     on_every_nth_instruction: Some(1), ..Default::default()
+/// };
+/// # let _ = triggers;
+/// # }
+/// ```
+pub struct HookTriggers {
+    /// Before a function call.
+    pub on_calls: bool,
+    /// When Lua returns from a function.
+    pub on_returns: bool,
```

Comment:
It might be a good idea to document that this does *not* include returns from tail-called functions:

> For call events, event can be LUA_HOOKCALL, the normal value, or LUA_HOOKTAILCALL, for a tail call; in this case, there will be no corresponding return event.